comms/core/src/protocol/rpc/client/pool.rs

```diff
@@ -90,34 +90,72 @@ where T: RpcPoolClient + From<RpcClient> + NamedProtocolService + Clone
+                },
                 None => match self.add_new_client_session().await {
-                    Ok(c) => c,
+                    Ok((c, len)) => {
+                        trace!(
+                            target: LOG_TARGET,
+                            "Added new RPC client session for connection '{}' ({} of {}, protocol: {:?})",
+                            peer_node_id, len, clients_capacity, protocol_id,
+                        );
+                        c
+                    },
                     // This is an edge case where the remote node does not have any further sessions available. This is
                     // gracefully handled by returning one of the existing used sessions.
-                    Err(RpcClientPoolError::NoMoreRemoteServerRpcSessions(val)) => self
-                        .get_least_used()
-                        .ok_or(RpcClientPoolError::NoMoreRemoteServerRpcSessions(val))?,
-                    Err(RpcClientPoolError::NoMoreRemoteClientRpcSessions(val)) => self
-                        .get_least_used()
-                        .ok_or(RpcClientPoolError::NoMoreRemoteClientRpcSessions(val))?,
+                    Err(RpcClientPoolError::NoMoreRemoteServerRpcSessions(val)) => {
+                        let c = self
+                            .get_least_used()
+                            .ok_or(RpcClientPoolError::NoMoreRemoteServerRpcSessions(val.clone()))?;
+                        trace!(
+                            target: LOG_TARGET,
+                            "used existing RPC client session for connection '{}', protocol: {:?} ({})",
+                            peer_node_id, protocol_id, RpcClientPoolError::NoMoreRemoteServerRpcSessions(val),
+                        );
+                        c
+                    },
+                    Err(RpcClientPoolError::NoMoreRemoteClientRpcSessions(val)) => {
+                        let c = self
+                            .get_least_used()
+                            .ok_or(RpcClientPoolError::NoMoreRemoteClientRpcSessions(val.clone()))?;
+                        trace!(
+                            target: LOG_TARGET,
+                            "used existing RPC client session for connection '{}', protocol: {:?} ({})",
+                            peer_node_id, protocol_id, RpcClientPoolError::NoMoreRemoteClientRpcSessions(val),
+                        );
+                        c
+                    },
                     Err(err) => {
                         return Err(err);
                     },
                 },
             };
 
             if !client.is_connected() {
+                trace!(
+                    target: LOG_TARGET,
+                    "RPC client for connection '{}' is not connected, pruning",
+                    peer_node_id
+                );
                 self.prune();
-                continue;
+                return Err(RpcClientPoolError::CouldNotObtainRpcConnection);
```

Comment:
This appears to return an (rather meaningless) error instead of pruning the disconnected session and returning another session that is connected. What is the reasoning here?

Reply:
At this stage the client should have been connected, as either an existing client with connection has been assigned, or a new client connection has been established. If it is not connected it is actually an error condition, and my reasoning is rather let the connecting peer decide what to do. This curcumvent a state where the pool tries to establish aconnection time after time if the server keeps on rejecting it.